### DISCUSSION TITLE
What's the proper way to spawn a task on current thread?

### DISCUSSION THREAD

**Discussion by RealXuChe** [tokio-rs/tokio#3947] [Q&A] [answered]
I want to create a task in the current thread because I don't want to take the synchronization overhead. I tried the following code but ran into problems :
```rust
LocalSet::new().run_until(async move {
    some code  // It seems that until it returns.......
}).await;
more code  // Code here will not be executed?
```
The task started with `run_until` blocks the program from continuing, and the behavior I want is close to the initial `spawn`, which will execute two tasks simultaneously.
```rust
spawn(async move {
    some code // This.....
});
other code // .....and this will be executed simultaneously.
```
What's the proper way to spawn a task on current thread?

**Comment by Darksonn** [comment, marked as answer]
If all you are using is the `LocalSet::run_until` method, you should consider simply using [`tokio::join!`](https://docs.rs/tokio/1/tokio/macro.join.html). Both halves of a `tokio::join!` are part of the same task, so they will be executed together on the same thread.

If you do want to use a `LocalSet`, then you should probably be spawning _both_ tasks on the `LocalSet` with its `spawn_local` method. You can then `.await` the `LocalSet` which completes once all tasks spawned on the set return.

**Reply by RealXuChe** [reply]
What I need is to have two tasks running on the same thread in the background, `LocalSet` and `join!` should return as soon as the task starts running, not wait until the task finishes running .

`spawn` has the behavior I want: it returns as soon as the task is started, but it can't execute the task on the same thread.

My code is roughly like this:
```rust
pub async fn bind<A: ToSocketAddrs>(addrs: A) -> std::io::Result<Self> {
    // bind `addrs`
    let (mut newconn_tx, newconn_rx) = channel::<NewConnInfo>(1);
    spawn(async move { // Have to spawn here because join! blocks
        let mut active_conn = Rc::new(RefCell::new(X));// `Rc` is not `Send+Sync`, so the two task have to run on current thread.
        let mut active_conn1 = active_conn.clone();
        let task_income = async move {
            loop {
                // wait for new connection and write active_conn
                // send new connection with `newconn_tx`.
            }
        };
        let task_revoke = async move {
            loop {
                // inspects the status of `active_conn1`, remove connection if needed.
            }
        };
        tokio::join!(task_income,task_revoke);
    });
    
    Ok(Self {
        newconn_rx, // can accept new connection via `newconn_rx` later
    })
}
```

**Reply by Darksonn** [reply]
You can use `tokio::task::spawn_local`, but this only works if the parent task is already running on a `LocalSet`. If the parent task is running as an ordinary `tokio::spawn` task, then this is not possible.

**Reply by Darksonn** [reply]
In principle, it would be correct if you use unsafe to smuggle the references in to the right place, something like this:
```Rust
struct JoinRef<'a, T> {
    value: &'a T,
}

unsafe impl<'a, T: Send> Send for JoinRef<'a, T> {}
```
```Rust
spawn(async move {
    let value = RefCell::new(X);
    let shared1 = JoinRef { value: &value };
    let shared2 = JoinRef { value: &value };
    let task_income = async move {
        loop {
            // use shared1
        }
    };
    let task_revoke = async move {
        loop {
            // use shared2
        }
    };
    tokio::join!(task_income,task_revoke);
});
```
But of course this is dangerous. If `shared1` and `shared2` ever somehow end up on two different threads, things can go pretty wrong, and the compiler wont catch it.

**Reply by RealXuChe** [reply]
Maybe it's possible to start another current-thread runtime in a new thread and run these tasks? 
But this doesn't sound practical enough, I should end up choosing the `unsafe` approach.
Thank you for your help!

**Reply by brandonros** [reply]
What's the equivalent of trying to join a JoinHandle from Tokio (async context) in non-async context?

```rust
fn main() {
  let thread_join_handle = std::thread::spawn(|| { ... })
  let result = thread_join_handle.join().unwrap()
}
```

kind of like this (non async main fn)

**Reply by Darksonn** [reply]
If you have a new unrelated question, please open a new thread for it.

**Reply by brandonros** [reply]
https://github.com/tokio-rs/tokio/discussions/3947#discussioncomment-2884271
https://github.com/tokio-rs/tokio/discussions/4563#discussioncomment-2869167
https://github.com/tokio-rs/tokio/discussions/3857#discussioncomment-2829652

**Reply by Darksonn** [reply]
If you don't know how to open a new thread, then you can do so at this page: https://github.com/tokio-rs/tokio/discussions/new
